source/contributing/implementation/datamanipulator.rst

```diff
@@ -0,0 +1,271 @@
+Here is a short example:
+
+.. code-block:: java
+
+  public class SpongeFlyingData extends AbstractBooleanData<FlyingData, ImmutableFlyingData> implements FlyingData {
+ /* your code here */
+  }
+
+The next step is to provide the generic ``DataManipulator`` that follows the generic of ``FlyingData``. In this case
+we use ``FlyingData.class`` in our ``super`` constructor, followed by the ``Key`` which identifies the recommended
+``value`` of the flying state and last but not least the ``Key`` itself: ``Keys.IS_FLYING``.
+
+.. code-block:: java
+
+ public SpongeFlyingData(boolean flying) {
+          super(FlyingData.class, flying, Keys.IS_FLYING);
+ }
+
+In most cases while implementing an abstract Manipulator you want to have two constructors:
+
+* One without arguments (no-args) which calls the second
+* and one single-argument constructor which is actually used for the provided ``value``
+
+Now, after the Manipulator is done you have to check if you are extending ``AbstractData`` or not.
+Remember that we decided to extend ``AbstractBooleanData`` above, hence now must call ``registerFieldGetter()`` and
+``registerFieldSetter()`` in your constructor. Don't forget the ``Field`` registration!
+
+.. code-block:: java
+
+ private void registerStuff() {
+      registerFieldGetter(Keys.IS_FLYING, new GetterFunction<Object>() {
+             @Override
+             public Value get() {
+                 return getflying();
+             }
+         });
+      registerFieldSetter(Keys.IS_FLYING, new SetterFunction<Object>() {
+              @Override
+              public void set(Object value) {
+                   setFlying(((Boolean) value).booleanValue());
+              }
+          });
+  }
+
+That's it. The ``DataManipulator`` should be done now.
+
+2. ImmutableDataManipulator
+===========================
+
+Just repeat the above mentioned steps for the ``ImmutableDataManipulator``.
+
+Now, after the ``Manipulators`` are done we have to register the ``KEY`` and finally implement the
+``DataProcessor`` and ``ValueProcessor``.
+
+3. Register the Key in the KeyRegistry
+======================================
+
+The next step is to register your ``KEYS`` in the ``KeyRegistry``:
+
+.. code-block:: java
+
+ public static void registerKeys() {
+      keyMap.put("is_flying", makeSingleKey(Boolean.class, Value.class, of("IsFlying")));
+ }
+
+
+4. Implement the DataProcessor
+==============================
+
+Next up is the ``DataProcessor``. A ``DataProcessor`` implements your ``DataManipulator`` to work semi-natively with
+Minecraft's objects. Unfortunately, since it's always easier to just directly implement all of the processing of a
+particular set of data in one place, we have ``DataProcessors`` to do just that.
+Currently, a ``DataProcessor`` has the following methods:
+
+.. code-block:: java
+
+ public interface DataProcessor<M extends DataManipulator<M, I>, I extends ImmutableDataManipulator<I, M>> {
```

Comment:
oh boy... BRAIN DUMP TIME!

So a few weeks ago, it was increasingly a problem that `DataProcessor`s were requiring far too much boilerplate code. What came out of it was the `AbstractDataProcessor` which included the default priority for Sponge provided data processors, and then further down the road, came about the `AbstractEntitySingleDataProcessor` and `AbstractEntityDataProcessor` which carefully determined that the data manipulator is only focused on an `net.minecraft.entity.Entity` and thankfully narrows down all of these methods into the following:

```java
    protected abstract M createManipulator();

    protected boolean supports(E entity) {
        return true;
    }

    protected abstract boolean set(E entity, T value);

    protected abstract Optional<T> getVal(E entity);

    protected abstract ImmutableValue<T> constructImmutableValue(T value);
```
And what ends up as a result of this generalization is:
```java
public class HealthDataProcessor extends AbstractEntityDataProcessor<EntityLivingBase, HealthData, ImmutableHealthData> {

    public HealthDataProcessor() {
        super(EntityLivingBase.class);
    }

    @Override
    protected HealthData createManipulator() {
        return new SpongeHealthData(20, 20);
    }

    @Override
    protected boolean doesDataExist(EntityLivingBase entity) {
        return true;
    }

    @Override
    protected boolean set(EntityLivingBase entity, Map<Key<?>, Object> keyValues) {
        entity.getEntityAttribute(SharedMonsterAttributes.maxHealth).setBaseValue(((Double) keyValues.get(Keys.MAX_HEALTH)).floatValue());
        entity.setHealth(((Double) keyValues.get(Keys.HEALTH)).floatValue());
        return true;
    }

    @Override
    protected Map<Key<?>, ?> getValues(EntityLivingBase entity) {
        final double health = entity.getHealth();
        final double maxHealth = entity.getMaxHealth();
        return ImmutableMap.<Key<?>, Object>of(Keys.HEALTH, health,
                                               Keys.MAX_HEALTH, maxHealth);
    }

    @Override
    public Optional<HealthData> fill(DataContainer container, HealthData healthData) {
        healthData.set(Keys.MAX_HEALTH, getData(container, Keys.MAX_HEALTH));
        healthData.set(Keys.HEALTH, getData(container, Keys.HEALTH));
        return Optional.of(healthData);
    }

    @Override
    public DataTransactionResult remove(DataHolder dataHolder) {
        return DataTransactionBuilder.builder().result(DataTransactionResult.Type.FAILURE).build();
    }

}
```

Ping me again later for more brain dumps.